core/query/query.mli

```diff
@@ -0,0 +1,104 @@
+open Utility;;
+
+type tag = int
+  [@@deriving show]
+
+type t =
+    [ `For of tag option * (Var.var * t) list * t list * t
+    | `If of t * t * t
+    | `Table of Value.table
+    | `Database of (Value.database * string)
+    | `Singleton of t | `Concat of t list
+    | `Record of t StringMap.t | `Project of t * string | `Erase of t * StringSet.t
+    | `Variant of string * t
+    | `XML of Value.xmlitem
+    | `Apply of string * t list
+    | `Closure of (Ir.var list * Ir.computation) * env
+    | `Primitive of string
+    | `Var of (Var.var * Types.datatype StringMap.t) | `Constant of Constant.constant ]
+and env = Value.env * t Env.Int.t
+  [@@deriving show]
+
+val unbox_xml : [> `XML of 'a ] -> 'a
+    ;;
+val unbox_pair : [> `Record of 'a Utility.StringMap.t ] -> 'a * 'a
+    ;;
+val unbox_list : ([> `Concat of 'a list | `Singleton of 'b ] as 'a) -> 'b list
+    ;;
+val unbox_string : [> `Concat of [> `Concat of 'a
+                                  | `Singleton of [> `Constant of [> `Char of char ] ] as 'b ]
+                list as 'a
+            | `Constant of [> `String of string ]
+            | `Singleton of 'b ] ->
+		string
+;;
+
+val used_database : t -> Value.database option
+    ;;
+
+val string_of_t : t -> string
+
+val type_of_expression : t -> Types.datatype
+    ;;
+
+val default_of_base_type : [> `Bool | `Char | `Float | `Int | `String ] ->
```

Comment:
I would create an alias of variant row, and name it, say, `base_type`.

Reply:
Makes sense. However I'm not entirely sure about why default_of_base_type has a catch-all case whose only purpose is to raise assert false.

Reply:
Changes made. I noticed there is a "Types.primitive" that's very similar to the new "Types.base_type", with extra cases for `XmlItem and `DB. I have chosen to keep the two types separate because, as the code changes, I would prefer to see a type error at compile time rather than a random assertion failure at runtime (raised e.g. by Query.default_of_base_type). If somebody feels these should be only one type, let me know.

Reply:
I think they should probably be separate types, since XmlItem and DB values are not possible to return in the result of a (flat SQL) query  It might be good to document the difference in the name, e.g. rename `Types.base_type` to `Types.query_primitive` or (eventually) `Sql.primitive` or something like that.

Reply:
I would probably implement the `base_type` (or whatever we call it) in the `Query` module to keep it local. If you want to keep it in the `Types` module, then you can avoid a bit of code duplication by
```
type base_type = [ `Bool | `Char | `Float | `Int | `String ]
type primitive_type = [ base_type | `XmlItem | `DB ]
```

Reply:
Yes, it makes sense to move this definition to Query. Changes made.